diff --git a/kdbg/gdbdriver.cpp b/kdbg/gdbdriver.cpp
index 9d006e8..2b5d4cf 100644
--- a/kdbg/gdbdriver.cpp
+++ b/kdbg/gdbdriver.cpp
@@ -2262,6 +2262,7 @@ uint GdbDriver::parseProgramStopped(const char* output, bool haveCoreFile,
     // "Program terminated with wignal SIGSEGV"
     // "Program received signal SIGINT" or other signal
     // "Breakpoint..."
+	   output[0] == '\n' ||
     // GDB since 7.3 prints
     // "[Inferior 1 (process 13400) exited normally]"
     // "[Inferior 1 (process 14698) exited with code 01]"
